java/ql/src/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -446,16 +452,14 @@ abstract class LocalCallContext extends TLocalFlowCallContext {
   abstract predicate validFor(Node n);
 }
 
-LocalCallContext getMatchingLocalCallContext(CallContext ctx) {
-  (
-    not ctx instanceof CallContextSpecificCall or
-    not exists(TSpecificLocalCall(ctx.(CallContextSpecificCall).getCall()))
-  ) and
-  exists(LocalCallContextAny l | result = l)
-  or
-  exists(LocalCallContextSpecificCall l |
-    ctx.(CallContextSpecificCall).getCall() = l.getCall() and result = l
-  )
+/**
+ * Gets a matching local call context given the call context and a node which is in
+ * the callable the call is targeting.
+ */
+LocalCallContext getMatchingLocalCallContext(CallContext ctx, Node n) {
+  if hasUnreachableNode(ctx.(CallContextSpecificCall).getCall(), n.getEnclosingCallable())
+  then result.(LocalCallContextSpecificCall).getCall() = ctx.(CallContextSpecificCall).getCall()
+  else result instanceof LocalCallContextAny
```

Comment:
As written, this will create a Cartesian product between `Node`s `n` and `CallContext`s `ctx`, when `not hasUnreachableNode(ctx.(CallContextSpecificCall).getCall(), n.getEnclosingCallable()` holds. One possibility is to inline the predicate (using `pragma[inline]`), @aschackmull : what do you think?